code/segmentedtickerplant/stplog.q

```diff
@@ -0,0 +1,303 @@
+      .stpps.pubclear[t]];
+  }each .stpps.t;
+ };
+
+// Standard batch mode - write to disk immediately, publish in batches
+upd[`defaultbatch]:{[t;x;now]
+  t insert x:.stplg.updtab[t] . (x;now);
+  `..loghandles[t] enlist(`upd;t;x);
+  // track tmp counts, and add these after publish
+  @[`.stplg.tmpmsgcount;t;+;1];
+  @[`.stplg.tmprowcount;t;+;count first x];
+ };
+
+zts[`defaultbatch]:{
+  // publish and clear all tables, increment counts
+  .stpps.pubclear[.stpps.t];
+  // after data has been published, updated the counts
+  .stplg.msgcount+:.stplg.tmpmsgcount;
+  .stplg.rowcount+:.stplg.tmprowcount;
+  // reset temp counts
+  .stplg.tmpmsgcount:.stplg.tmprowcount:()!();
+ };
+
+// Immediate mode - publish and write immediately
+upd[`immediate]:{[t;x;now]
+  x:updtab[t] . (x;now);
+  `..loghandles[t] enlist(`upd;t;x);
+  x:$[0h>type last x;enlist;flip] .stpps.tabcols[t]!x;
+  @[`.stplg.msgcount;t;+;1];
+  @[`.stplg.rowcount;t;+;count x];
+  .stpps.pub[t;x]
+ };
+
+zts[`immediate]:{}
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+// Functions to obtain logs for client replay ////////////////////////////////////////
+// replaylog called from client-side, returns nested list of logcounts and lognames
+replaylog:{[t]
+  getlogs[replayperiod][t]
+ }
+
+// alternative replay allows for 'pass through logging'
+// if SCTP not producing logs, subscribers replay from STP log files
+if[.sctp.loggingmode=`parent;
+  replaylog:{[t]
+    .sctp.tph (`.stplg.replaylog; t)
+    }
+  ]
+
+getlogs:enlist[`]!enlist ()
+
+// If replayperiod set to `period, only replay logs for current logging period
+getlogs[`period]:{[t]
+  distinct flip (.stplg.msgcount;exec tbl!logname from `..currlog where tbl in t)@\:t
+ };
+
+// If replayperiod set to `day, replay all of today's logs
+getlogs[`day]:{[t]
+  // set the msgcount to 0Wj for all logs which have closed
+  lnames:select seq,tbls,logname,msgcount:0Wj from .stpm.metatable where any each tbls in\: t;
+  // Meta table does not store counts for live logs, so these are populated here
+  lnames:update msgcount:sum each .stplg.msgcount[tbls] from lnames where seq=.stplg.i;
+  flip value exec `long$msgcount,logname from lnames
+ };
+
+//////////////////////////////////////////////////////////////////////////////////////
+
+// Open log for a single table at start of logging period
+openlog:{[multilog;dir;tab;p]
+  lname:logname[multilog][dir;tab;p];
+  .lg.o[`openlog;"opening logfile: ",string lname];
+  h:$[(notexists:not type key lname)or null h0:exec first handle from `..currlog where logname=lname;
+    [if[notexists;.[lname;();:;()]];hopen lname];
+    h0
+  ];
+  `..currlog upsert (tab;lname;h);
+ };
+
+// Error log for failed updates in error mode
+openlogerr:{[dir]
+  lname:.[{hsym`$string[x],"/",string[y],(raze string"dv"$(.z.p+.eodtime.dailyadj)) except".:"};(dir;`$"_" sv string .proc.procname,errorlogname);{.lg.e[`openlogerr;"failed to make error log: ",x]}];
```

Comment:
do we add the dailyadj to the timestamp on the other log files ? 
(I don't know, can you check)
This log file should use the timestamping function created above

Reply:
We do add the dailyadj, yes
We'll get that function added!